pkg/storegateway/batch_series.go

```diff
@@ -22,165 +21,45 @@ import (
 	"github.com/grafana/mimir/pkg/util/pool"
 )
 
-type blockSeriesChunkRefsSetsIterator struct {
-	ctx      context.Context
+type postingsSetsIterator struct {
 	postings []storage.SeriesRef
 
 	batchSize                  int
 	currentBatchPostingsOffset int
```

Comment:
The logic around `currentBatchPostingsOffset` is a bit tricky (e.g. you have to initialise it to `-batchSize`). I think the logic would be cleaner you change this field to `nextBatchPostingsOffset`, so that it gets initialized to 0 and it's the offset of the next batch. You shouldn't need the offset of the current one, given the current batch is already stored in `currentBatch`.